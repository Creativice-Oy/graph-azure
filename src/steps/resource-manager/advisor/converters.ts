import {
  Entity,
  createIntegrationEntity,
  convertProperties,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { AdvisorEntities } from './constants';
import { ResourceRecommendationBase } from '@azure/arm-advisor/esm/models';

function maybeConvertToNumber(maybeNumber: string | undefined) {
  if (!isNaN(parseFloat(maybeNumber || 'NaN'))) return parseFloat(maybeNumber!);
}

export function createRecommendationEntity(
  webLinker: AzureWebLinker,
  data: ResourceRecommendationBase,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: AdvisorEntities.RECOMMENDATION._type,
        _class: AdvisorEntities.RECOMMENDATION._class,
        id: data.id,
        name: data.name,
        category: data.category,
        impact: data.impact,
        impactedField: data.impactedField,
        impactedValue: data.impactedValue,
        recommendationTypeId: data.recommendationTypeId,
        shortDescriptionProblem: data.shortDescription?.problem,
        shortDescriptionSolution: data.shortDescription?.solution,
        assessmentKey: data.extendedProperties?.assessmentKey,
        score: maybeConvertToNumber(data.extendedProperties?.score),
        resourceId: data.resourceMetadata?.resourceId,
        source: data.resourceMetadata?.source,
        open: true,
        severity: data.impact,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
