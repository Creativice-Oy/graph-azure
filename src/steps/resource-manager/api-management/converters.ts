import {
  Entity,
  createIntegrationEntity,
  convertProperties,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { ApiManagementEntities } from './constants';
import {
  ApiManagementServiceResource,
  ApiContract,
} from '@azure/arm-apimanagement/esm/models';

export function createApiManagementServiceEntity(
  webLinker: AzureWebLinker,
  data: ApiManagementServiceResource,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: ApiManagementEntities.SERVICE._type,
        _class: ApiManagementEntities.SERVICE._class,
        id: data.id,
        name: data.name,
        public: !!data.publicIPAddresses?.length,
        function: ['api-gateway'],
        category: ['application'],
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}

export function createApiManagementApiEntity(
  webLinker: AzureWebLinker,
  data: ApiContract,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: ApiManagementEntities.API._type,
        _class: ApiManagementEntities.API._class,
        id: data.id,
        name: data.name,
        address: data.path,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
