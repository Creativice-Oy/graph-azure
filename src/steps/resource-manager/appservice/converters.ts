import {
  Entity,
  createIntegrationEntity,
  StepEntityMetadata,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { AppServicePlan, Site } from '@azure/arm-appservice/esm/models';
import { AppServiceEntities } from './constants';

export function createAppEntity(
  webLinker: AzureWebLinker,
  data: Site,
  metadata: StepEntityMetadata,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: data.id as string,
        _type: metadata._type,
        _class: metadata._class,
        id: data.id,
        name: data.name,
        type: data.type,
        kind: data.kind?.split(','),
        location: data.location,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}

export function createAppServicePlanEntity(
  webLinker: AzureWebLinker,
  data: AppServicePlan,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: data.id as string,
        _type: AppServiceEntities.APP_SERVICE_PLAN._type,
        _class: AppServiceEntities.APP_SERVICE_PLAN._class,
        id: data.id,
        name: data.name,
        type: data.type,
        kind: data.kind?.split(','),
        location: data.location,
        'sku.name': data.sku?.name,
        'sku.tier': data.sku?.tier,
        'sku.size': data.sku?.size,
        'sku.family': data.sku?.family,
        'sku.capacity': data.sku?.capacity,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
