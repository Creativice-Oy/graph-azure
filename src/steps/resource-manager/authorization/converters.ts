import {
  RoleDefinition,
  Permission,
} from '@azure/arm-authorization/esm/models';
import { convertProperties, Entity } from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import {
  ROLE_DEFINITION_ENTITY_CLASS,
  ROLE_DEFINITION_ENTITY_TYPE,
} from './constants';

export function convertPermissionsToRulesList(
  permissions: Permission[] | undefined,
): string[] {
  const rules: string[] = [];
  // [{ actions: ['*'], notActions: [], dataActions: [], notDataActions: [] }]
  if (permissions) {
    // { actions: ['*'], notActions: [], dataActions: [], notDataActions [] }
    for (const permission of permissions) {
      // actions || notActions || dataActions || notDataActions
      for (const actionKey of Object.keys(permission)) {
        // '*'
        for (const action of permission[actionKey]) {
          // 'actions = *'
          rules.push(`${actionKey} = ${action}`);
        }
      }
    }
  }
  return rules;
}

export function createRoleDefinitionEntity(
  webLinker: AzureWebLinker,
  data: RoleDefinition,
): Entity {
  const entity = {
    ...convertProperties(data),
    _key: data.id as string,
    _type: ROLE_DEFINITION_ENTITY_TYPE,
    _class: ROLE_DEFINITION_ENTITY_CLASS,
    _rawData: [{ name: 'default', rawData: data }],
    displayName: data.roleName,
    description: data.description,
    rules: convertPermissionsToRulesList(data.permissions),
    webLink: webLinker.portalResourceUrl(data.id),
  };
  return entity;
}
