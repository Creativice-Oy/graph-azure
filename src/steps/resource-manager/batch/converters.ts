import {
  Entity,
  createIntegrationEntity,
  convertProperties,
} from '@jupiterone/integration-sdk-core';
import { AzureWebLinker } from '../../../azure';
import { BatchEntities } from './constants';
import { BatchAccount, Pool } from '@azure/arm-batch/esm/models';

export function createBatchAccountEntity(
  webLinker: AzureWebLinker,
  data: BatchAccount,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: BatchEntities.BATCH_ACCOUNT._type,
        _class: BatchEntities.BATCH_ACCOUNT._class,
        id: data.id,
        name: data.name,
        category: ['infrastructure'],
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}

export function createBatchPoolEntity(
  webLinker: AzureWebLinker,
  data: Pool,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: BatchEntities.BATCH_POOL._type,
        _class: BatchEntities.BATCH_POOL._class,
        id: data.id,
        name: data.name,
        webLink: webLinker.portalResourceUrl(data.id),
        type: data.type,
      },
    },
  });
}

export function createBatchApplicationEntity(
  webLinker: AzureWebLinker,
  data: Pool,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: BatchEntities.BATCH_APPLICATION._type,
        _class: BatchEntities.BATCH_APPLICATION._class,
        id: data.id,
        name: data.displayName,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
