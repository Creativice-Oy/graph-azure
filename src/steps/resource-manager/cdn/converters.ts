import {
  Entity,
  createIntegrationEntity,
  convertProperties,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { CdnEntities } from './constants';
import { Profile, Endpoint } from '@azure/arm-cdn/esm/models';

export function createCdnProfileEntity(
  webLinker: AzureWebLinker,
  data: Profile,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: CdnEntities.PROFILE._type,
        _class: CdnEntities.PROFILE._class,
        id: data.id,
        name: data.name,
        category: ['infrastructure'],
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}

export function createCdnEndpointEntity(
  webLinker: AzureWebLinker,
  data: Endpoint,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: CdnEntities.ENDPOINT._type,
        _class: CdnEntities.ENDPOINT._class,
        id: data.id,
        name: data.name,
        category: ['data'],
        function: ['content-distribution'],
        public: true,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
