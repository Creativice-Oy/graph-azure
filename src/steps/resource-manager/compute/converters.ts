import { Disk, Image, VirtualMachine } from '@azure/arm-compute/esm/models';
import {
  assignTags,
  convertProperties,
  createDirectRelationship,
  Entity,
  getTime,
  Relationship,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { resourceGroupName } from '../../../azure/utils';
import {
  DISK_ENTITY_CLASS,
  DISK_ENTITY_TYPE,
  VIRTUAL_MACHINE_ENTITY_CLASS,
  VIRTUAL_MACHINE_ENTITY_TYPE,
  VIRTUAL_MACHINE_IMAGE_ENTITY_CLASS,
  VIRTUAL_MACHINE_IMAGE_ENTITY_TYPE,
} from './constants';

export function createVirtualMachineEntity(
  webLinker: AzureWebLinker,
  data: VirtualMachine,
): Entity {
  const osProperties = {};
  if (data.storageProfile) {
    Object.assign(osProperties, {
      platform: data.storageProfile.osDisk?.osType?.toLowerCase(),
      osName: data.storageProfile.imageReference?.offer,
      osVersion: data.storageProfile.imageReference?.exactVersion,
    });
  }
  if (data.osProfile) {
    Object.assign(osProperties, {
      adminUser: data.osProfile.adminUsername,
      disablePasswordAuthentication:
        data.osProfile.linuxConfiguration?.disablePasswordAuthentication,
      enableAutomaticUpdates:
        data.osProfile.windowsConfiguration?.enableAutomaticUpdates,
      timeZone: data.osProfile.windowsConfiguration?.timeZone,
    });
  }
  const entity = {
    ...convertProperties(data),
    ...osProperties,
    _key: data.id as string,
    _type: VIRTUAL_MACHINE_ENTITY_TYPE,
    _class: VIRTUAL_MACHINE_ENTITY_CLASS,
    _rawData: [{ name: 'default', rawData: data }],
    displayName: data.name,
    vmId: data.vmId,
    type: data.type,
    region: data.location,
    resourceGroup: resourceGroupName(data.id),
    state: data.provisioningState,
    vmSize: data.hardwareProfile && data.hardwareProfile.vmSize,
    webLink: webLinker.portalResourceUrl(data.id),
  };

  assignTags(entity, data.tags);

  return entity;
}

export function createDiskEntity(
  webLinker: AzureWebLinker,
  data: Disk,
): Entity {
  const entity = {
    ...convertProperties(data),
    _key: data.id as string,
    _type: DISK_ENTITY_TYPE,
    _class: DISK_ENTITY_CLASS,
    _rawData: [{ name: 'default', rawData: data }],
    displayName: data.name,
    region: data.location,
    resourceGroup: resourceGroupName(data.id),
    createdOn: getTime(data.timeCreated),
    webLink: webLinker.portalResourceUrl(data.id),
    encrypted: !!data.encryption?.type,
    state: data.diskState?.toLowerCase(),
    attached: data.diskState === 'Attached',
  };

  assignTags(entity, data.tags);

  return entity;
}

export function createImageEntity(
  webLinker: AzureWebLinker,
  data: Image,
): Entity {
  const entity = {
    ...convertProperties(data),
    _key: data.id as string,
    _type: VIRTUAL_MACHINE_IMAGE_ENTITY_TYPE,
    _class: VIRTUAL_MACHINE_IMAGE_ENTITY_CLASS,
    _rawData: [{ name: 'default', rawData: data }],
    displayName: data.name,
    region: data.location,
    resourceGroup: resourceGroupName(data.id),
    webLink: webLinker.portalResourceUrl(data.id),
  };

  assignTags(entity, data.tags);

  return entity;
}

export function createVirtualMachineDiskRelationships(
  vm: VirtualMachine,
): Relationship[] {
  const relationships: Relationship[] = [];

  if (vm.storageProfile) {
    if (vm.storageProfile.osDisk?.managedDisk?.id) {
      relationships.push(
        createDirectRelationship({
          _class: RelationshipClass.USES,
          fromKey: vm.id as string,
          fromType: VIRTUAL_MACHINE_ENTITY_TYPE,
          toKey: vm.storageProfile.osDisk.managedDisk.id,
          toType: DISK_ENTITY_TYPE,
          properties: {
            osDisk: true,
          },
        }),
      );
    }

    for (const disk of vm.storageProfile.dataDisks || []) {
      if (disk.managedDisk?.id) {
        relationships.push(
          createDirectRelationship({
            _class: RelationshipClass.USES,
            fromKey: vm.id as string,
            fromType: VIRTUAL_MACHINE_ENTITY_TYPE,
            toKey: disk.managedDisk.id,
            toType: DISK_ENTITY_TYPE,
            properties: {
              dataDisk: true,
            },
          }),
        );
      }
    }
  }

  return relationships;
}
