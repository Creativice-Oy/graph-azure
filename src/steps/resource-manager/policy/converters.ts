import { AzureWebLinker } from '../../../azure';
import {
  createIntegrationEntity,
  Entity,
} from '@jupiterone/integration-sdk-core';

import { flatten } from './helpers';
import { PolicyEntities } from './constants';
import { PolicyAssignment } from '@azure/arm-policy/esm/models';

export function createPolicyAssignmentEntity(
  webLinker: AzureWebLinker,
  data: PolicyAssignment,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...flatten(data),
        _key: data.id,
        _type: PolicyEntities.POLICY_ASSIGNMENT._type,
        _class: PolicyEntities.POLICY_ASSIGNMENT._class,
        id: data.id,
        name: data.displayName,
        webLink: webLinker.portalResourceUrl(data.id),
        scope: data.scope,
        policyDefinitionId: data.policyDefinitionId,
      },
    },
  });
}
