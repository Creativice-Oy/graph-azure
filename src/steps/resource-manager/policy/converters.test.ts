import { createAzureWebLinker } from '../../../azure';
import { createPolicyAssignmentEntity } from './converters';
import { PolicyAssignment } from '@azure/arm-policy/esm/models';

const webLinker = createAzureWebLinker('something.onmicrosoft.com');

describe('createPolicyAssignmentEntity', () => {
  test('properties transferred', () => {
    const data: PolicyAssignment = {
      displayName: 'Enforce resource naming rules',
      description:
        'Force resource names to begin with given DeptA and end with -LC',
      metadata: {
        assignedBy: 'Someone Special',
      },
      policyDefinitionId:
        '/subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2/providers/Microsoft.Authorization/policyDefinitions/ResourceNaming',
      notScopes: [],
      parameters: {
        prefix: {
          value: 'DeptA',
        },
        suffix: {
          value: '-LC',
        },
      },
      enforcementMode: 'Default',
      scope: 'subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2',
      id:
        '/subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2/providers/Microsoft.Authorization/policyAssignments/EnforceNaming',
      type: 'Microsoft.Authorization/policyAssignments',
      name: 'EnforceNaming',
    };

    const policyAssignmentEntity = createPolicyAssignmentEntity(
      webLinker,
      data,
    );

    expect(policyAssignmentEntity).toMatchSnapshot();
    expect(policyAssignmentEntity).toMatchGraphObjectSchema({
      _class: ['ControlPolicy'],
      schema: {
        additionalProperties: true,
        properties: {
          scope: {
            const: 'subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2',
          },
          policyDefinitionId: {
            const:
              '/subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2/providers/Microsoft.Authorization/policyDefinitions/ResourceNaming',
          },
        },
      },
    });
  });
});
