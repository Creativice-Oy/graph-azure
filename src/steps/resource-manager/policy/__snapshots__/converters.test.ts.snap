// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPolicyAssignmentEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "ControlPolicy",
  ],
  "_key": "/subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2/providers/Microsoft.Authorization/policyAssignments/EnforceNaming",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "description": "Force resource names to begin with given DeptA and end with -LC",
        "displayName": "Enforce resource naming rules",
        "enforcementMode": "Default",
        "id": "/subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2/providers/Microsoft.Authorization/policyAssignments/EnforceNaming",
        "metadata": Object {
          "assignedBy": "Someone Special",
        },
        "name": "EnforceNaming",
        "notScopes": Array [],
        "parameters": Object {
          "prefix": Object {
            "value": "DeptA",
          },
          "suffix": Object {
            "value": "-LC",
          },
        },
        "policyDefinitionId": "/subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2/providers/Microsoft.Authorization/policyDefinitions/ResourceNaming",
        "scope": "subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2",
        "type": "Microsoft.Authorization/policyAssignments",
      },
    },
  ],
  "_type": "azure_policy_assignment",
  "createdOn": undefined,
  "description": "Force resource names to begin with given DeptA and end with -LC",
  "displayName": "Enforce resource naming rules",
  "enforcementMode": "Default",
  "id": "/subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2/providers/Microsoft.Authorization/policyAssignments/EnforceNaming",
  "metadata.assignedBy": "Someone Special",
  "name": "Enforce resource naming rules",
  "parameters.prefix": "DeptA",
  "parameters.suffix": "-LC",
  "policyDefinitionId": "/subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2/providers/Microsoft.Authorization/policyDefinitions/ResourceNaming",
  "scope": "subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2",
  "type": "Microsoft.Authorization/policyAssignments",
  "webLink": "https://portal.azure.com/#@something.onmicrosoft.com/resource/subscriptions/ae640e6b-ba3e-4256-9d62-2993eecfa6f2/providers/Microsoft.Authorization/policyAssignments/EnforceNaming",
}
`;
